src/main/java/org/reaktivity/nukleus/tls/internal/stream/TlsServerFactory.java

```diff
@@ -412,66 +404,71 @@ private int decodeClientHello(
             {
-                if (tlsRecordInfo.type() == TlsContentType.HANDSHAKE.value())
+                assert newTlsEngine.getHandshakeSession() instanceof ExtendedSSLSession;
+
+                final List<SNIServerName> requestedServerNames = ((ExtendedSSLSession) newTlsEngine.getHandshakeSession())
+                    .getRequestedServerNames();
+                byte[] tlsHostname = !requestedServerNames.isEmpty() ? requestedServerNames.get(0).getEncoded() : null;
+
+                final MessagePredicate filter = (t, b, i, l) ->
                 {
-                    final TlsHandshakeFW tlsHandshake = tlsHandshakeRO.tryWrap(buffer, tlsFragmentOffset, tlsRecordLimit);
+                    final TlsRouteExFW routeExFW = wrapRouteEx.apply(t, b, i, l);
 
-                    if (tlsHandshake != null && tlsHandshake.type() == TlsHandshakeType.CLIENT_HELLO.value())
-                    {
-                        final TlsClientHelloFW tlsClientHello = tlsHandshake.body().get(tlsClientHelloRO::tryWrap);
+                    final String8FW hostname1 = routeExFW.hostname();
+                    final byte[] hostname = new byte[hostname1.length()];
+                    hostname1.buffer().getBytes(hostname1.offset() + hostname1.fieldSizeLength(),
+                        hostname, 0, hostname1.length());
 
-                        if (tlsClientHello != null)
-                        {
-                            server.onDecodeClientHello(traceId, authorization, tlsClientHello);
-                        }
-                    }
-                }
+                    return tlsHostname != null && Arrays.equals(tlsHostname, hostname);
+                };
 
-                if (server.tlsEngine == null)
-                {
-                    counters.serverDecodeNoClientHello.getAsLong();
-                }
-                else
-                {
-                    try
-                    {
-                        server.tlsEngine.beginHandshake();
-                        server.decoder = decodeHandshake;
-                    }
-                    catch (SSLException | RuntimeException ex)
-                    {
-                        server.tlsEngine = null;
-                    }
-                }
+                final RouteFW newRoute = router.resolve(server.routeId, authorization, filter, wrapRoute);
+                final TlsRouteExFW newRouteEx = newRoute.extension().get(tlsRouteExRO.get()::tryWrap);
 
-                if (server.tlsEngine == null)
-                {
-                    server.cleanupNetwork(traceId);
-                    server.decoder = decodeIgnoreAll;
-                }
+                return newRouteEx.protocol().asString();
+            });
+
+            server.tlsEngine = newTlsEngine;
+            server.tlsStoreInfo = newTlsStoreInfo;
+        }
+
+        if (server.tlsEngine == null)
+        {
+            counters.serverDecodeNoClientHello.getAsLong();
```

Comment:
We can remove these counters that rely on decoding the `ClientHello` directly as we don't need them any more.